File: src/main/java/JetFactory.java
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.jet.Jet;
import com.hazelcast.jet.JetInstance;

/**
 * @author Robin Duda
 *
 */
public class JetFactory {
    private static transient JetInstance jet;

    /**
     *
     * @return
     */
    public static JetInstance jetInstance() {
        if (jet == null) {
            synchronized (JetProcessBuilder.class) {
                if (jet == null) {
                    jet = Jet.newJetInstance();
                }
            }
        }
        return jet;
    }

    /**
     *
     * @return
     */
    public static HazelcastInstance hazelInstance() {
        return jetInstance().getHazelcastInstance();
    }
}

File: src/main/kotlin/com/github/codingchili/process/impl/JetFactory.kt
package com.github.codingchili.process.impl

import com.hazelcast.config.Config
import com.hazelcast.core.HazelcastInstance
import com.hazelcast.jet.Jet
import com.hazelcast.jet.JetInstance
import com.hazelcast.jet.config.JetConfig
import java.util.*

/**
 * @author Robin Duda
 *
 * Factory for creating instances of the Jet distributed stream processing engine.
 */
object JetFactory {
    @Transient
    private var jet: JetInstance? = null
    @Transient
    private val id = UUID.randomUUID().toString().substring(0,7)

    /**
     * @return the Jet instance if one exists, otherwise one is created.
     */
    fun jetInstance(): JetInstance {
        if (jet == null) {
            synchronized(javaClass) {
                if (jet == null) {
                    jet = Jet.newJetInstance(JetConfig().setHazelcastConfig(Config().setInstanceName("jet_instance_$id")))
                }
            }
        }
        return jet!!
    }

    /**
     * @return the hazelcast instance associated with the current Jet instance.
     */
    fun hazelInstance(): HazelcastInstance {
        return jetInstance().hazelcastInstance
    }
}
